lib/matplotlib/axes/_axes.py

```diff
@@ -2604,6 +2605,123 @@ def barh(self, y, width, height=0.8, left=None, *, align="center",
+
+        Returns
+        -------
+        annotations
+            A list of `.Text` instances for the labels.
+        """
+        def extend_with_none(iterable, length):
+            return list(iterable) + [None] * (length - len(iterable))
+
+        def sign(x):
+            return 1 if x >= 0 else -1
+
+        cbook._check_in_list(['edge', 'center'], position=position)
+
+        bars = container.patches
+        errorbar = container.errorbar
+        orientation = container.orientation
+
+        N = len(bars)
+        if errorbar:
+            lines = errorbar.lines
+            barlinecols = lines[2]
+            barlinecol = barlinecols[0]
+            segments = barlinecol.get_segments()
+            errs = extend_with_none(segments, length=N)
+        else:
+            errs = extend_with_none([], length=N)
+        caps = extend_with_none(captions, length=N)
+
+        annotations = []
+        for bar, err, cap in zip(bars, errs, caps):
+
+            (x0, y0), (x1, y1) = bar.get_bbox().get_points()
+            xc, yc = (x0 + x1) / 2, (y0 + y1) / 2
+
+            if orientation == "vertical":
+                extrema = max(y0, y1) if yc >= 0 else min(y0, y1)
+                length = abs(y0 - y1)
+            elif orientation == "horizontal":
+                extrema = max(x0, x1) if xc >= 0 else min(x0, x1)
+                length = abs(x0 - x1)
+
+            if err is None:
+                endpt = extrema
+            elif orientation == "vertical":
+                endpt = err[:, 1].max() if yc >= 0 else err[:, 1].min()
+            elif orientation == "horizontal":
+                endpt = err[:, 0].max() if xc >= 0 else err[:, 0].min()
+
+            if position == "center":
+                value = sign(extrema) * length
+            elif position == "edge":
+                value = extrema
+
+            if position == "center":
+                xy = xc, yc
+            elif position == "edge" and orientation == "vertical":
+                xy = xc, endpt
+            elif position == "edge" and orientation == "horizontal":
+                xy = endpt, yc
+
+            if orientation == "vertical":
+                xytext = shifting, sign(extrema) * padding
+            else:
+                xytext = sign(extrema) * padding, shifting
+
+            if position == "center":
+                ha, va = "center", "center"
+            elif position == "edge" and orientation == "vertical" and yc >= 0:
+                ha, va = "center", "bottom"
+            elif position == "edge" and orientation == "vertical" and yc < 0:
+                ha, va = "center", "top"
+            elif (
+                position == "edge" and orientation == "horizontal" and xc >= 0
+            ):
+                ha, va = "left", "center"
+            elif position == "edge" and orientation == "horizontal" and xc < 0:
+                ha, va = "right", "center"
+
+            annotation = self.annotate(cap or fmt % value, xy, xytext,
```

Comment:
This will ignore user input of an empty string, probably better to check for `None`-ness?